Add logic in master service to optimize performance and retain detailed logging for critical cluster operations.

sumitasr opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
Add logic in master service to optimize performance and retain detailed logging for critical cluster operations.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/pull/14795#pullrequestreview-2363701033

### Testing

Modified the check for generating short summary from 1000 to 1 in local environment to test the logging.

[2024-10-22T13:44:32,280][DEBUG][o.o.c.s.MasterService    ] [runTask-0] took [0s] to notify listeners on successful publication of cluster state (version: 6, uuid: _4hvlyf0RA2ulV2jtjIsKg) for [Tasks batched with key: org.opensearch.cluster.action.shard.ShardStateAction, count:2 and sample tasks: shard-started StartedShardEntry{shardId [[sample-index2][1]], allocationId [wZdnPxXmTQ-87FSRVNA8CQ], primary term [1], message [master {runTask-0}{d12XZzhcSiG1aGyfi-eUpw}{s36FgDsCS6Git2YUlAXRzw}{127.0.0.1}{127.0.0.1:9300}{dimr}{testattr=test, shard_indexing_pressure_enabled=true} marked shard as initializing, but shard state is [POST_RECOVERY], mark shard as started]}[StartedShardEntry{shardId [[sample-index2][1]], allocationId [wZdnPxXmTQ-87FSRVNA8CQ], primary term [1], message [master {runTask-0}{d12XZzhcSiG1aGyfi-eUpw}{s36FgDsCS6Git2YUlAXRzw}{127.0.0.1}{127.0.0.1:9300}{dimr}{testattr=test, shard_indexing_pressure_enabled=true} marked shard as initializing, but shard state is [POST_RECOVERY], mark shard as started]}], shard-started StartedShardEntry{shardId [[sample-index2][1]], allocationId [wZdnPxXmTQ-87FSRVNA8CQ], primary term [1], message [after new shard recovery]}[StartedShardEntry{shardId [[sample-index2][1]], allocationId [wZdnPxXmTQ-87FSRVNA8CQ], primary term [1], message [after new shard recovery]}]]


[2024-10-22T13:44:32,057][DEBUG][o.o.c.s.MasterService    ] [runTask-0] took [0s] to notify listeners on successful publication of cluster state (version: 4, uuid: FeOWukKcR5qoPMRZMIlqKA) for [Tasks batched with key: org.opensearch.cluster.metadata.MetadataCreateIndexService, count:1 and sample tasks: create-index [sample-index2], cause [api]]


[2024-10-22T13:44:32,282][DEBUG][o.o.c.s.MasterService    ] [runTask-0] took [2ms] to compute cluster state update for [Tasks batched with key: org.opensearch.cluster.routing.BatchedRerouteService, count:1 and sample tasks: cluster_reroute(reroute after starting shards)]



Node Join

[2024-10-25T12:51:37,680][INFO ][o.o.c.s.MasterService ] [data1] Tasks batched with key: org.opensearch.cluster.coordination.JoinHelper, count:1 and sample tasks: node-join[{data1}{bUKIJ8aDR6yJftPHtNrMVg}{-chi7PdPRjS2TwqxUnTkLQ}{127.0.0.1}{127.0.0.1:9301}{dmr}{shard_indexing_pressure_enabled=true} join existing leader], term: 2, version: 7, delta: added {{data1}{bUKIJ8aDR6yJftPHtNrMVg}{-chi7PdPRjS2TwqxUnTkLQ}{127.0.0.1}{127.0.0.1:9301}{dmr}{shard_indexing_pressure_enabled=true}}

Node Left

[2024-10-25T12:52:05,089][INFO ][o.o.c.s.MasterService ] [data1] Tasks batched with key: org.opensearch.cluster.coordination.NodeRemovalClusterStateTaskExecutor@78a30062, count:1 and sample tasks: node-left[{data1}{bUKIJ8aDR6yJftPHtNrMVg}{-chi7PdPRjS2TwqxUnTkLQ}{127.0.0.1}{127.0.0.1:9301}{dmr}{shard_indexing_pressure_enabled=true} reason: disconnected], term: 2, version: 8, delta: removed {{data1}{bUKIJ8aDR6yJftPHtNrMVg}{-chi7PdPRjS2TwqxUnTkLQ}{127.0.0.1}{127.0.0.1:9301}{dmr}{shard_indexing_pressure_enabled=true}}

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Bukhtawar reviewed (commented):
Thanks can we add a test for `node-left` logs please

shwetathareja commented on server/src/main/java/org/opensearch/cluster/service/MasterService.java:
```diff
@@ -299,8 +299,15 @@ public static boolean assertNotMasterUpdateThread(String reason) {
-        final String shortSummary = taskInputs.taskSummaryGenerator.apply(false);
+        final String longSummary;
+        final String shortSummary;
+        if (taskInputs.updateTasks.size() <= 1000) {
```
@Bukhtawar should we keep this limit to 100 or 50, that should be good enough to give a picture of tasks being processed.

shwetathareja commented on server/src/main/java/org/opensearch/cluster/service/TaskBatcher.java:
```diff
@@ -195,25 +195,37 @@ void runIfNotProcessed(BatchedTask updateTask) {
                     if (longSummaryRequired == null || !longSummaryRequired) {
-                        return buildShortSummary(updateTask.batchingKey, toExecute.size());
+                        final List<BatchedTask> sampleTasks = toExecute.stream()
+                            .limit(Math.min(1000, toExecute.size()))
```
let keep this as 100

sumitasr replied:
ack

sumitasr replied:
@Bukhtawar Will 100 tasks summary work for node-join and node-left use-cases?

sumitasr replied:
We will keep it to 1000 for now.

shwetathareja commented on server/src/main/java/org/opensearch/cluster/service/MasterService.java:
```diff
@@ -299,8 +299,15 @@ public static boolean assertNotMasterUpdateThread(String reason) {
+            longSummary = taskInputs.taskSummaryGenerator.apply(true);
+            shortSummary = longSummary;
+        } else {
+            longSummary = logger.isTraceEnabled() ? taskInputs.taskSummaryGenerator.apply(true) : "";
+            shortSummary = taskInputs.taskSummaryGenerator.apply(false);
```
it will compute twice now. 

shouldnt this be:

```
if(logger.isTraceEnabled()) {
    summary = taskInputs.taskSummaryGenerator.apply(true);
} else {
    summary = taskInputs.taskSummaryGenerator.apply(false);
}
```

Now, for shortSummary, it already takes care of logging n tasks (which is 1000 here). We can remove multiple if check on tracelog and simply use summary everywhere?

sumitasr replied:
We were logging INFO at few places with short summary irrespective of trace log level and only with long summary where trace log was required.

For now refactoring as suggested to keep code simple.

shwetathareja replied:
what was the being logged before from those info logs before you made long/short summary change?

sumitasr replied:
long summary only.

sumitasr commented:
> Thanks can we add a test for `node-left` logs please


Node Join

[2024-10-25T12:51:37,680][INFO ][o.o.c.s.MasterService    ] [data1] Tasks batched with key: org.opensearch.cluster.coordination.JoinHelper, count:1 and sample tasks: node-join[{data1}{bUKIJ8aDR6yJftPHtNrMVg}{-chi7PdPRjS2TwqxUnTkLQ}{127.0.0.1}{127.0.0.1:9301}{dmr}{shard_indexing_pressure_enabled=true} join existing leader], term: 2, version: 7, delta: added {{data1}{bUKIJ8aDR6yJftPHtNrMVg}{-chi7PdPRjS2TwqxUnTkLQ}{127.0.0.1}{127.0.0.1:9301}{dmr}{shard_indexing_pressure_enabled=true}}

Node Left

[2024-10-25T12:52:05,089][INFO ][o.o.c.s.MasterService    ] [data1] Tasks batched with key: org.opensearch.cluster.coordination.NodeRemovalClusterStateTaskExecutor@78a30062, count:1 and sample tasks: node-left[{data1}{bUKIJ8aDR6yJftPHtNrMVg}{-chi7PdPRjS2TwqxUnTkLQ}{127.0.0.1}{127.0.0.1:9301}{dmr}{shard_indexing_pressure_enabled=true} reason: disconnected], term: 2, version: 8, delta: removed {{data1}{bUKIJ8aDR6yJftPHtNrMVg}{-chi7PdPRjS2TwqxUnTkLQ}{127.0.0.1}{127.0.0.1:9301}{dmr}{shard_indexing_pressure_enabled=true}}

shwetathareja commented:
> Tasks batched with key: org.opensearch.cluster.coordination.JoinHelper, count:1 and
@sumitasr should we append this test in the end?

sumitasr commented:
> > Tasks batched with key: org.opensearch.cluster.coordination.JoinHelper, count:1 and
> > @sumitasr should we append this test in the end?

I think it will be hard to understand the log when lets say we have 100 tasks. This key kind of groups them together. I think it is helpful to have this in beginning.

sumitasr commented:
> The backport to `2.x` failed:
> 
> ```
> The process '/usr/bin/git' failed with exit code 128
> ```
> 
> To backport manually, run these commands in your terminal:
> 
> ```shell
> # Navigate to the root of your repository
> cd $(git rev-parse --show-toplevel)
> # Fetch latest updates from GitHub
> git fetch
> # Create a new working tree
> git worktree add ../.worktrees/OpenSearch/backport-2.x 2.x
> # Navigate to the new working tree
> pushd ../.worktrees/OpenSearch/backport-2.x
> # Create a new branch
> git switch --create backport/backport-16421-to-2.x
> # Cherry-pick the merged commit of this pull request and resolve the conflicts
> git cherry-pick -x --mainline 1 6f1b59e54bec41d40772f8571c7b65d4b523f8b1
> # Push it to GitHub
> git push --set-upstream origin backport/backport-16421-to-2.x
> # Go back to the original working tree
> popd
> # Delete the working tree
> git worktree remove ../.worktrees/OpenSearch/backport-2.x
> ```
> 
> Then, create a pull request where the `base` branch is `2.x` and the `compare`/`head` branch is `backport/backport-16421-to-2.x`.

Raised backport PR https://github.com/opensearch-project/OpenSearch/pull/16493
